lib/backend/pgbk/background.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pgbk
+
+import (
+	"context"
+	"encoding/hex"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgtype/zeronull"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	pgcommon "github.com/gravitational/teleport/lib/backend/pgbk/common"
+	"github.com/gravitational/teleport/lib/defaults"
+)
+
+func (b *Backend) backgroundExpiry(ctx context.Context) {
+	defer b.log.Info("Exited expiry loop.")
+
+	for ctx.Err() == nil {
+		// "DELETE FROM kv WHERE expires <= now()" but more complicated: logical
+		// decoding can become really really slow if a transaction is big enough
+		// to spill on disk - max_changes_in_memory (4096) changes before
+		// Postgres 13, or logical_decoding_work_mem (64MiB) bytes of total size
+		// in Postgres 13 and later; thankfully, we can just limit our
+		// transactions to a small-ish number of affected rows (1000 seems to
+		// work ok) as we don't need atomicity for this; we run a tight loop
+		// here because it could be possible to have more than ExpiryBatchSize
+		// new items expire every ExpiryInterval, so we could end up not ever
+		// catching up
+		for i := 0; i < backend.DefaultRangeLimit/b.cfg.ExpiryBatchSize; i++ {
+			t0 := time.Now()
+			deleted, err := pgcommon.RetryIdempotent(ctx, b.log, func() (int64, error) {
+				// TODO(espadolini): try other alternatives:
+				// - `DELETE FROM kv WHERE key IN (SELECT ...)`, might be faster
+				//   or slower depending on the batch size
+				// - `WITH to_delete AS (SELECT ...) DELETE FROM ... USING to_delete WHERE key = to_delete.key`
```

Comment:
This sounds the most "natural" to me. Have you also though about creating a temp table?

Reply:
The performance is certainly good enough with all these options; I've tweaked it slightly in 8fd585fa67b4b610d342bc202e0f40a0285cde5a to select keys in expiration order, which seems to have slightly better performance when combined with the LIMIT (it uses the index on expires rather than doing a seq scan which might take a while before it gets enough rows - the planner can't know how many will be selected). We can change it later if the behavior under heavy real load shows that the expiry deletion ends up forcing too many serialization retries; for instance, I suspect that grabbing a connection, doing a `CREATE TEMP TABLE to_delete AS SELECT key FROM kv WHERE (item is expired)` in read-only, deferred mode and then later doing a `DELETE FROM kv WHERE (item is expired and key is in to_delete)` might be better with regards to messing the least with the rest of the backend operations.